src/robot/output/listeners.py

```diff
@@ -15,298 +15,591 @@
-                                                   instantiate_with_args=args)
-        return listener, name
+class LibraryListeners(Listeners):
 
-    def _get_version(self, listener):
-        try:
-            version = int(listener.ROBOT_LISTENER_API_VERSION)
-            if version not in (2, 3):
-                raise ValueError
-        except AttributeError:
-            raise DataError("Listener '%s' does not have mandatory "
-                            "'ROBOT_LISTENER_API_VERSION' attribute."
-                            % self.name)
-        except (ValueError, TypeError):
-            raise DataError("Listener '%s' uses unsupported API version '%s'."
-                            % (self.name, listener.ROBOT_LISTENER_API_VERSION))
-        return version
-
-    @classmethod
-    def import_listeners(cls, listeners, method_names, prefix=None,
-                         raise_on_error=False):
-        imported = []
-        for listener in listeners:
-            try:
-                imported.append(cls(listener, method_names, prefix))
-            except DataError as err:
-                name = listener if is_string(listener) else type_name(listener)
-                msg = "Taking listener '%s' into use failed: %s" % (name, err)
-                if raise_on_error:
-                    raise DataError(msg)
-                LOGGER.error(msg)
-        return imported
-
-
-class ListenerAdapter(LoggerApi):
-
-    def __init__(self, listener):
-        self.listener = listener
+    def __init__(self, log_level='INFO'):
+        super().__init__([], log_level)
+        self._listener_stack = []
+
+    @property
+    def listeners(self):
+        return self._listener_stack[-1] if self._listener_stack else []
+
+    def new_suite_scope(self):
+        self._listener_stack.append([])
+
+    def discard_suite_scope(self):
+        self._listener_stack.pop()
 
     def register(self, listeners, library):
-        self.listener.register(listeners, library)
+        listeners = import_listeners(listeners, library=library)
+        self._listener_stack[-1].extend(listeners)
+
+    def close(self):
+        pass
 
     def unregister(self, library, close=False):
-        self.listener.unregister(library, close)
+        if close:
+            for listener in [li for li in self.listeners if li.library is library]:
+                listener.close()
+        listeners = [listener for listener in self._listener_stack[-1] if listener.library is not library]
+        self._listener_stack[-1] = listeners
 
-    def new_suite_scope(self):
-        self.listener.new_suite_scope()
 
-    def discard_suite_scope(self):
-        self.listener.discard_suite_scope()
+class ListenerFacade(LoggerApi):
 
-    def message(self, message: 'model.Message'):
-        self.listener.message(message)
+    def __init__(self, listener, name, prefix=''):
+        self.listener = listener
+        self.name = name
+        self.prefix = prefix
```

Comment:
We only use `prefix` when a leading underscore is allowed. Changing `prefix=''` to `allow_leading_underscore=False` could be a good idea.